src/coreclr/jit/lower.cpp

```diff
@@ -2249,6 +2230,297 @@ GenTree* Lowering::LowerTailCallViaJitHelper(GenTreeCall* call, GenTree* callTar
     return result;
 }
 
+//------------------------------------------------------------------------
+// LowerCFGCall: Potentially lower a call to use control-flow guard. This
+// expands indirect calls into either a validate+call sequence or to a dispatch
+// helper taking the original target in a special register.
+//
+// Arguments:
+//    call         -  The call node
+//
+void Lowering::LowerCFGCall(GenTreeCall* call)
+{
+    if (call->IsHelperCall(comp, CORINFO_HELP_VALIDATE_INDIRECT_CALL) ||
+        call->IsHelperCall(comp, CORINFO_HELP_DISPATCH_INDIRECT_CALL))
+    {
+        return;
+    }
+
+    GenTree* callTarget = call->gtCallType == CT_INDIRECT ? call->gtCallAddr : call->gtControlExpr;
+    if ((callTarget == nullptr) || callTarget->IsIntegralConst())
+    {
+        // This is a direct call, no CFG check is necessary.
+        return;
+    }
+
+    CFGCallKind cfgKind = call->GetCFGCallKind();
+
+    switch (cfgKind)
+    {
+        case CFGCallKind::ValidateAndCall:
+        {
+            // To safely apply CFG we need to generate a very specific pattern:
+            // in particular, it is a safety issue to allow the JIT to reload
+            // the call target from memory between calling
+            // CORINFO_HELP_VALIDATE_INDIRECT_CALL and the target. This is
+            // something that would easily occur in debug codegen if we
+            // produced high-level IR. Instead we will use a GT_PHYSREG node
+            // to get the target back from the register that contains the target.
+            //
+            // Additionally, the validator does not preserve all arg registers,
+            // so we have to move all GT_PUTARG_REG nodes that would otherwise
+            // be trashed ahead. The JIT also has an internal invariant that
+            // once GT_PUTARG nodes start to appear in LIR, the call is coming
+            // up. To avoid breaking this invariant we move _all_ GT_PUTARG
+            // nodes (in particular, GC info reporting relies on this).
+            //
+            // To sum up, we end up transforming
+            //
+            // ta... = <early args>
+            // tb... = <late args>
```

Comment:
Another dumb question, what is the difference between early and late args?

Reply:
Definitely not a dumb question, and I'm not even sure I understand it myself. :-)

Here's what I understand: `GenTreeCall` has three different fields that store nodes related to args: `gtCallThisArg`, `gtCallArgs` and `gtCallLateArgs`. The two former fields are what I call "early args" here while the latter is the "late args".
Before morph, the latter is not used, the early args just store the operand nodes for the call. Life is good.
`fgMorphArgs`, as part of morph, is responsible for setting calls up for the optimization/backend to handle, and among other things handles ABI details of how to pass the argument in either registers or on the stack. 

This comes with a constraint: we cannot start placing values in registers (and for non-x86, on the stack) before we have evaluated the arguments, if the arguments contain calls. To handle these cases morph will replace the early arguments with evaluation nodes that evaluate the argument into a temp, and then the temp will be put in the late args. In some cases no evaluation is necessary (e.g. for a simple constant) and the operand will end up either in the early or late args, depending on whether it is going on the stack or in registers.

So for a call like `GetThis().M(Arg1(), 10, Arg2())` we end up transforming something like
```scala
GT_CALL
  gtCallThisArg: GT_CALL GetThis
  gtCallArgs[0]: GT_CALL Arg1
  gtCallArgs[1]: GT_CNS_INT 10
  gtCallArgs[2]: GT_CALL Arg2
```
into
```scala
GT_CALL
  gtCallThisArg: GT_ASG tmp0 (GT_CALL GetThis)
  gtCallArgs[0]: GT_ASG tmp1 (GT_CALL Arg1)
  gtCallArgs[1]: GT_ARGPLACE
  gtCallArgs[2]: GT_ASG tmp2 (GT_CALL Arg2)
  gtCallLateArgs[0]: GT_LCL_VAR tmp0
  gtCallLateArgs[1]: GT_LCL_VAR tmp1
  gtCallLateArgs[2]: GT_CNS_INT 10
  gtCallLateArgs[3]: GT_LCL_VAR tmp2
```

For stack arguments we might not need any entry in `gtCallLateArgs`, so there isn't a 1-1 correspondence between args and late args.

The code in lower is then responsible for creating the nodes to actually place these args into the stack/registers. Morph has pretty much done all the work at this point so this is just creating `GT_PUTARG_REG` and `GT_PUTARG_STK` nodes and inserting them after the corresponding args in the early/late args lists. These particular nodes is what this comment is talking about: it moves them further down than they would normally be, since it needs to insert a call very late (after evaluation of the call target that comes after the late args nodes).